File: src/tca/reference/sorting/selection_sort.py
from tca.core.instrumentation import Metrics, Trace, make_probe


def selection_sort(
    values,
    metrics: Metrics | None = None,
    trace: Trace | None = None,
) -> None:
    probe = make_probe(metrics, trace)

    for index_i in range(len(values) - 1):  # i=(1)..(n-1)
        marker = index_i  # m=i

        probe.event(
            "select_minimum",
            indices=(marker,),
            values=(values[marker],),
        )  # mínimo atual m=i

        for index_j in range(index_i + 1, len(values)):  # j=(i+1)..(n)
            if probe.lt(
                values[index_j],
                values[marker],
                indices=(index_j, marker),
                roles=("current", "marker"),
            ):  # se x_j < x_m
                marker = index_j  # m=j

                probe.event(
                    "select_minimum",
                    indices=(marker,),
                    values=(values[marker],),
                )  # novo mínimo m=j

        probe.swap(values, index_i, marker)  # swap(x_i, x_m)

File: cpp/src/algorithms/sorting/selection_sort.cpp
#include "tca/algorithms/sorting/selection_sort.hpp"

#include <utility>

#include "tca/core/instrumentation/direct_probe.hpp"
#include "tca/core/instrumentation/probe.hpp"

namespace {

// snippet:start selection-sort
template <typename ProbeType>
void selection_sort_impl(std::span<double> values, ProbeType& probe) {
    for (std::size_t index_i = 0; index_i < values.size(); ++index_i) { // i=(0)..(n-1)
        std::size_t marker = index_i;                                   // m=i

        for (std::size_t index_j = index_i + 1; index_j < values.size();
             ++index_j) {                                    // j=(i+1)..(n)
            if (probe.lt(values[index_j], values[marker])) { // se xj < xm
                marker = index_j;                            // m=j
            }
        }

        probe.swap(values, index_i, marker); // swap(x_i, x_m)
    }
}
// snippet:end selection-sort

} // namespace

namespace tca::algorithms {

void selection_sort(std::span<double> values) {
    tca::instrumentation::DirectProbe probe;
    selection_sort_impl(values, probe);
}

void selection_sort(std::span<double> values, tca::instrumentation::Metrics& metrics) {
    tca::instrumentation::Probe probe(metrics);
    selection_sort_impl(values, probe);
}

} // namespace tca::algorithms
